pkg/daemon/containerrecreate/crr_daemon_controller.go

```diff
@@ -0,0 +1,471 @@
+			if duration < maxExpectationWaitDuration {
+				notSatisfied = true
+				continue
+			}
+			klog.Warningf("Wait for CRR %s/%s resourceVersion expectation over %v", crr.Namespace, crr.Name, duration)
+			resourceVersionExpectation.Delete(crr)
+		}
+
+		switch crr.Status.Phase {
+		case "":
+			if err := c.updateCRRPhase(crr, appsv1alpha1.ContainerRecreateRequestPending); err != nil {
+				klog.Errorf("Failed to update CRR %s/%s status to Pending: %v", crr.Namespace, crr.Name, err)
+				return nil, err
+			}
+		case appsv1alpha1.ContainerRecreateRequestPending:
+			if firstPending == nil {
+				firstPending = crr
+			}
+		case appsv1alpha1.ContainerRecreateRequestRecreating:
+			if firstRecreating == nil {
+				firstRecreating = crr
+			}
+		}
+	}
+	if notSatisfied {
+		return nil, nil
+	}
+	if firstRecreating != nil {
+		return firstRecreating, nil
+	}
+	if firstPending != nil {
+		return firstPending, nil
+	}
+	return nil, nil
+}
+
+func (c *Controller) manage(crr *appsv1alpha1.ContainerRecreateRequest) error {
+	runtimeManager, err := c.newRuntimeManager(c.runtimeFactory, crr)
+	if err != nil {
+		klog.Errorf("Failed to find runtime service for %s/%s: %v", crr.Namespace, crr.Name, err)
+		return c.completeCRRStatus(crr, fmt.Sprintf("failed to find runtime service: %v", err))
+	}
+
+	pod := convertCRRToPod(crr)
+
+	podStatus, err := runtimeManager.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed to GetPodStatus %s/%s with uid %s: %v", pod.Namespace, pod.Name, pod.UID, err)
+	}
+	klog.V(5).Infof("CRR %s/%s for Pod %s GetPodStatus: %v", crr.Namespace, crr.Name, pod.Name, util.DumpJSON(podStatus))
+
+	newCRRContainerRecreateStates := getCurrentCRRContainersRecreateStates(crr, podStatus)
+	if !reflect.DeepEqual(crr.Status.ContainerRecreateStates, newCRRContainerRecreateStates) {
+		return c.patchCRRContainerRecreateStates(crr, newCRRContainerRecreateStates)
+	}
+
+	var completedCount int
+	for i := range newCRRContainerRecreateStates {
+		state := &newCRRContainerRecreateStates[i]
+		switch state.Phase {
+		case appsv1alpha1.ContainerRecreateRequestSucceeded:
+			completedCount++
+			continue
+		case appsv1alpha1.ContainerRecreateRequestFailed:
+			completedCount++
+			if crr.Spec.Strategy.FailurePolicy == appsv1alpha1.ContainerRecreateRequestFailurePolicyIgnore {
+				continue
+			}
+			return c.completeCRRStatus(crr, "")
+		case appsv1alpha1.ContainerRecreateRequestPending, appsv1alpha1.ContainerRecreateRequestRecreating:
+		}
+
+		if state.Phase == appsv1alpha1.ContainerRecreateRequestRecreating {
+			if crr.Spec.Strategy.OrderedRecreate {
+				break
+			}
+			continue
+		}
+
+		kubeContainerStatus := podStatus.FindContainerStatusByName(state.Name)
+		if kubeContainerStatus == nil {
+			break
+		}
+
+		msg := fmt.Sprintf("Stopping container %s by ContainerRecreateRequest %s", state.Name, crr.Name)
+		err := runtimeManager.KillContainer(pod, kubeContainerStatus.ID, state.Name, msg, nil)
```

Comment:
is it possible the killing action hangs here ?

Reply:
the `TerminationGracePeriodSeconds` in mock pod will be the killing deadline